File: sdk/cpp/examples/gold_dataset.cpp
/**
 * Gold dataset reference test for the GeoTrace C++ SDK.
 *
 * Reads the CSV fixtures in tests/fixtures/gold_dataset/, builds a .gtd file,
 * then verifies the round-trip.  Run from the repository root:
 *
 *   ./sdk/cpp/build/gold/examples/gold_dataset
 */

#include <geotrace/geotrace.hpp>

#include <algorithm>
#include <array>
#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <exception>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <iterator>
#include <map>
#include <optional>
#include <set>
#include <sstream>
#include <stdexcept>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

namespace fs = std::filesystem;

namespace {

std::string rtrim(std::string text) {
    while (!text.empty() && (text.back() == '\r' || text.back() == '\n' || text.back() == ' ')) {
        text.pop_back();
    }
    return text;
}

std::vector<std::string> split(const std::string &line, char delim) {
    std::vector<std::string> cols;
    std::istringstream stream(line);
    std::string col;
    while (std::getline(stream, col, delim)) {
        cols.push_back(std::move(col));
    }
    // std::getline drops the trailing empty field for lines that end with the delimiter.
    if (!line.empty() && line.back() == delim) {
        cols.emplace_back();
    }
    return cols;
}

std::vector<std::string> split_csv(const std::string &line) {
    return split(line, ',');
}

// The first N comma-separated fields of `line`, or `std::nullopt` when it holds fewer.
template <std::size_t N>
std::optional<std::array<std::string, N>> split_csv_fields(const std::string &line) {
    auto cols = split_csv(line);
    if (cols.size() < N) {
        return std::nullopt;
    }
    std::array<std::string, N> fields;
    std::move(cols.begin(), cols.begin() + static_cast<std::ptrdiff_t>(N), fields.begin());
    return fields;
}

bool is_leap(int year) noexcept {
    return (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0);
}

int month_days(int month, int year) noexcept {
    static constexpr std::array<int, 12> dom = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    return (month == 2 && is_leap(year)) ? 29 : dom.at(static_cast<std::size_t>(month - 1));
}

/* Parse "YYYY-MM-DDTHH:MM:SS+HH:MM" to a Timestamp, or `std::nullopt` on failure. */
std::optional<geotrace::Timestamp> parse_ts(const std::string &text) {
    if (text.size() < 19) {
        return std::nullopt;
    }
    try {
        auto year = std::stoi(text.substr(0, 4));
        auto month = std::stoi(text.substr(5, 2));
        auto day = std::stoi(text.substr(8, 2));
        auto hour = std::stoi(text.substr(11, 2));
        auto minute = std::stoi(text.substr(14, 2));
        auto second = std::stoi(text.substr(17, 2));
        // Optional fractional seconds (".ffffff"), kept as microseconds.
        std::size_t pos = 19;
        long frac_us = 0;
        if (pos < text.size() && text.at(pos) == '.') {
            ++pos;
            std::string digits;
            while (pos < text.size() && text.at(pos) >= '0' && text.at(pos) <= '9') {
                digits.push_back(text.at(pos));
                ++pos;
            }
            digits.resize(6, '0'); // pad / truncate to microseconds
            frac_us = std::stol(digits);
        }
        // Optional timezone offset ("+HH:MM" / "-HH:MM").
        char sign = '+';
        int tz_hours = 0;
        int tz_minutes = 0;
        if (pos < text.size() && (text.at(pos) == '+' || text.at(pos) == '-')) {
            sign = text.at(pos);
            if (pos + 6 <= text.size()) {
                tz_hours = std::stoi(text.substr(pos + 1, 2));
                tz_minutes = std::stoi(text.substr(pos + 4, 2));
            }
        }
        long days = 0;
        for (int y = 1970; y < year; y++) {
            days += is_leap(y) ? 366 : 365;
        }
        for (int m = 1; m < month; m++) {
            days += month_days(m, year);
        }
        days += day - 1;
        long secs = (days * 86400L) + (hour * 3600L) + (minute * 60L) + second;
        const long tz_seconds = ((static_cast<long>(tz_hours) * 60L) + tz_minutes) * 60L;
        secs += (sign == '-') ? tz_seconds : -tz_seconds;
        const std::uint64_t micros =
            (static_cast<std::uint64_t>(secs) * 1000000ULL) + static_cast<std::uint64_t>(frac_us);
        return geotrace::Timestamp::from_micros(micros);
    } catch (const std::exception &) {
        return std::nullopt;
    }
}

geotrace::Constellation parse_constellation(const std::string &name) {
    if (name == "gps") {
        return geotrace::Constellation::Gps;
    }
    if (name == "glonass") {
        return geotrace::Constellation::Glonass;
    }
    if (name == "galileo") {
        return geotrace::Constellation::Galileo;
    }
    if (name == "beidou") {
        return geotrace::Constellation::Beidou;
    }
    throw std::invalid_argument("unknown constellation: " + name);
}

geotrace::MarkerIcon parse_icon(const std::string &name) {
    if (name.empty() || name == "auto") {
        return geotrace::MarkerIcon::Auto;
    }
    if (name == "pin") {
        return geotrace::MarkerIcon::Pin;
    }
    if (name == "cross") {
        return geotrace::MarkerIcon::Cross;
    }
    if (name == "circle") {
        return geotrace::MarkerIcon::Circle;
    }
    if (name == "lightning") {
        return geotrace::MarkerIcon::Lightning;
    }
    if (name == "warning") {
        return geotrace::MarkerIcon::Warning;
    }
    if (name == "error") {
        return geotrace::MarkerIcon::Error;
    }
    if (name == "check") {
        return geotrace::MarkerIcon::Check;
    }
    if (name == "satellite") {
        return geotrace::MarkerIcon::Satellite;
    }
    if (name == "satellite_lost") {
        return geotrace::MarkerIcon::SatelliteLost;
    }
    if (name == "gear") {
        return geotrace::MarkerIcon::Gear;
    }
    if (name == "refresh") {
        return geotrace::MarkerIcon::Refresh;
    }
    if (name == "download") {
        return geotrace::MarkerIcon::Download;
    }
    if (name == "upload") {
        return geotrace::MarkerIcon::Upload;
    }
    if (name == "wrench") {
        return geotrace::MarkerIcon::Wrench;
    }
    return geotrace::MarkerIcon::Auto;
}

std::optional<double> parse_opt_double(const std::string &text) {
    if (text.empty()) {
        return std::nullopt;
    }
    try {
        std::size_t pos = 0;
        double value = std::stod(text, &pos);
        return (pos > 0) ? std::optional<double>{value} : std::nullopt;
    } catch (const std::exception &) {
        return std::nullopt;
    }
}

struct SatRow {
    std::string gps_time;
    std::string sys_time;
    geotrace::Satellite sat;
};

geotrace::FixTime required_fix_time(const geotrace::RecordedFixTimestamps &recorded,
                                    const std::string &source) {
    const auto time = geotrace::FixTime::from_recorded(recorded);
    if (!time) {
        throw geotrace::IoError(source + " has no timestamp");
    }
    return *time;
}

std::ifstream open_csv(const fs::path &base, const std::string &name) {
    auto path = base / name;
    std::ifstream file(path);
    if (!file.is_open()) {
        throw geotrace::IoError("cannot open: " + path.string());
    }
    return file;
}

void load_meta(geotrace::FileBuilder &builder, const fs::path &base) {
    auto file = open_csv(base, "meta.csv");
    std::string line;
    std::getline(file, line); // skip header
    if (!std::getline(file, line)) {
        throw geotrace::IoError("meta.csv: missing data row");
    }
    auto fields = split_csv_fields<5>(rtrim(std::move(line)));
    if (!fields) {
        throw geotrace::IoError("meta.csv: need 5 columns");
    }
    const auto &[title, device, notes, identity, travel_mode_name] = *fields;
    auto travel_mode = geotrace::travel_mode_from_name(travel_mode_name);
    if (!travel_mode) {
        throw geotrace::IoError("meta.csv: unknown travel mode: " + travel_mode_name);
    }
    builder.title(title).device(device).notes(notes).identity(identity).travel_mode(*travel_mode);
}

void load_event_styles(geotrace::FileBuilder &builder, const fs::path &base) {
    auto file = open_csv(base, "event_styles.csv");
    std::string line;
    std::getline(file, line); // skip header
    while (std::getline(file, line)) {
        line = rtrim(std::move(line));
        if (line.empty()) {
            continue;
        }
        auto fields = split_csv_fields<3>(line);
        if (!fields) {
            continue;
        }
        const auto &[variant_path, icon, color] = *fields;
        builder.add_event_marker_style(
            geotrace::EventMarkerStyle{variant_path, parse_icon(icon), color});
    }
}

std::vector<SatRow> load_satellites(const fs::path &base) {
    auto file = open_csv(base, "satellites.csv");
    std::string line;
    std::getline(file, line); // skip header
    std::vector<SatRow> rows;
    while (std::getline(file, line)) {
        line = rtrim(std::move(line));
        if (line.empty()) {
            continue;
        }
        auto fields = split_csv_fields<8>(line);
        if (!fields) {
            continue;
        }
        const auto &[gps_time, sys_time, constellation, prn, in_fix, elevation, azimuth, snr] =
            *fields;
        rows.push_back(SatRow{
            gps_time,
            sys_time,
            geotrace::Satellite{
                parse_constellation(constellation),
                static_cast<std::uint32_t>(std::stoul(prn)),
                in_fix == "true",
                parse_opt_double(elevation),
                parse_opt_double(azimuth),
                parse_opt_double(snr),
            },
        });
    }
    return rows;
}

void load_fixes(geotrace::FileBuilder &builder, const fs::path &base,
                const std::vector<SatRow> &sats) {
    auto file = open_csv(base, "fixes.csv");
    std::set<std::pair<std::string, std::string>> fix_times;
    std::string line;
    std::getline(file, line); // skip header
    while (std::getline(file, line)) {
        line = rtrim(std::move(line));
        if (line.empty()) {
            continue;
        }
        auto fields = split_csv_fields<8>(line);
        if (!fields) {
            continue;
        }
        [[maybe_unused]] const auto &[track_id, gps_time, sys_time, lat, lon, heading_deg,
                                      speed_kmh, eph_m] = *fields;

        geotrace::RecordedFixTimestamps recorded{};
        recorded.gps_time = parse_ts(gps_time);
        recorded.sys_time = parse_ts(sys_time);
        const auto time = required_fix_time(recorded, "fixes.csv row " + line);

        auto hdg = parse_opt_double(heading_deg);
        auto kmh = parse_opt_double(speed_kmh);

        builder.add(geotrace::NavFix{
            time,
            geotrace::Angle::degrees(std::stod(lat)),
            geotrace::Angle::degrees(std::stod(lon)),
            hdg ? std::optional{geotrace::Angle::degrees(*hdg)} : std::nullopt,
            kmh ? std::optional{geotrace::Velocity::kmh(*kmh)} : std::nullopt,
            parse_opt_double(eph_m),
        });

        std::vector<geotrace::Satellite> tracked;
        for (const auto &row : sats) {
            if (row.gps_time == gps_time && row.sys_time == sys_time) {
                tracked.push_back(row.sat);
            }
        }
        if (!tracked.empty()) {
            builder.add(geotrace::SatelliteReport{time, std::move(tracked)});
        }

        fix_times.emplace(gps_time, sys_time);
    }

    // Reports at a time no fix row holds. The builder gives each one a ghost fix.
    std::map<std::pair<std::string, std::string>, std::vector<geotrace::Satellite>> orphans;
    for (const auto &row : sats) {
        if (fix_times.find({row.gps_time, row.sys_time}) == fix_times.end()) {
            orphans[{row.gps_time, row.sys_time}].push_back(row.sat);
        }
    }
    for (const auto &[times, tracked] : orphans) {
        geotrace::RecordedFixTimestamps recorded{};
        recorded.gps_time = parse_ts(times.first);
        recorded.sys_time = parse_ts(times.second);

        const auto time = required_fix_time(recorded, "satellites.csv row (" + times.first + ", " +
                                                          times.second + ")");
        builder.add(geotrace::SatelliteReport{time, tracked});
    }
}

void load_markers(geotrace::FileBuilder &builder, const fs::path &base) {
    auto file = open_csv(base, "markers.csv");
    std::string line;
    std::getline(file, line); // skip header
    while (std::getline(file, line)) {
        line = rtrim(std::move(line));
        if (line.empty()) {
            continue;
        }
        auto fields = split_csv_fields<3>(line);
        if (!fields) {
            continue;
        }
        const auto &[time, label, icon] = *fields;
        auto timestamp = parse_ts(time);
        if (!timestamp) {
            throw geotrace::IoError("markers.csv: missing timestamp");
        }
        builder.add(geotrace::Annotation{*timestamp, label, parse_icon(icon)});
    }
}

void load_events(geotrace::FileBuilder &builder, const fs::path &base) {
    auto file = open_csv(base, "events.csv");
    std::string line;
    std::getline(file, line); // skip header
    while (std::getline(file, line)) {
        line = rtrim(std::move(line));
        if (line.empty()) {
            continue;
        }
        auto fields = split_csv_fields<3>(line);
        if (!fields) {
            continue;
        }
        const auto &[sys_time, variant_path, annotation] = *fields;
        auto timestamp = parse_ts(sys_time);
        if (!timestamp) {
            throw geotrace::IoError("events.csv: missing sys_time");
        }
        builder.add(geotrace::EventMarker{variant_path, *timestamp, annotation});
    }
}

// Returns the channel named in the row. Appends a new channel from the row's
// unit, period, description and component columns when no earlier row named it.
geotrace::Channel &channel_for_row(std::vector<geotrace::Channel> &channels,
                                   const std::array<std::string, 7> &fields) {
    const auto &[name, unit, period_deg, description, components, time, values] = fields;
    for (auto &existing : channels) {
        if (existing.name == name) {
            return existing;
        }
    }
    geotrace::Channel channel;
    channel.name = name;
    if (!unit.empty()) {
        channel.unit = geotrace::ChannelUnit::parse_recognized(unit);
    }
    if (!period_deg.empty()) {
        channel.period = geotrace::Angle::degrees(std::stod(period_deg));
    }
    channel.description = description;
    if (!components.empty()) {
        channel.components = split(components, ';');
    }
    channels.push_back(std::move(channel));
    return channels.back();
}

void load_channels(geotrace::FileBuilder &builder, const fs::path &base) {
    auto file = open_csv(base, "channels.csv");
    std::string line;
    std::getline(file, line); // skip header

    // One Channel per name, built up across its rows (each row is one sample).
    std::vector<geotrace::Channel> channels;
    while (std::getline(file, line)) {
        line = rtrim(std::move(line));
        if (line.empty()) {
            continue;
        }
        auto fields = split_csv_fields<7>(line);
        if (!fields) {
            continue;
        }
        [[maybe_unused]] const auto &[name, unit, period_deg, description, components, time,
                                      values] = *fields;

        geotrace::Channel &channel = channel_for_row(channels, *fields);
        auto timestamp = parse_ts(time);
        if (!timestamp) {
            throw geotrace::IoError("channels.csv: invalid timestamp");
        }
        channel.times.push_back(*timestamp);
        for (const auto &value : split(values, ';')) {
            channel.values.push_back(std::stod(value));
        }
    }

    for (const auto &channel : channels) {
        builder.add_channel(channel);
    }
}

void verify_counts(const geotrace::NavFile &file) {
    auto check = [](bool cond, const char *msg) {
        if (!cond) {
            throw std::runtime_error(std::string("FAIL: ") + msg);
        }
    };

    check(file.title().find("Gold Dataset") != std::string_view::npos, "title missing");
    check(file.device().find("Synthetic Generator") != std::string_view::npos, "device missing");
    check(file.notes().find("cross-SDK") != std::string_view::npos, "notes missing");
    check(file.identity() == "gold-standard-v2", "identity wrong");
    check(file.travel_mode() == "bicycle", "travel mode wrong");

    auto nav_points = file.nav_point_count();
    if (nav_points != 200) {
        throw std::runtime_error("expected 200 nav points, got " + std::to_string(nav_points));
    }

    std::size_t anti = 0;
    for (std::size_t i = 0; i < nav_points; i++) {
        auto point = file.nav_point(i);
        if (point.lon.as_degrees() > 179.9 || point.lon.as_degrees() < -179.9) {
            anti++;
        }
    }
    if (anti != 11) {
        throw std::runtime_error("expected 11 antimeridian pts, got " + std::to_string(anti));
    }

    auto event_markers = file.event_marker_count();
    if (event_markers != 7) {
        throw std::runtime_error("expected 7 event markers, got " + std::to_string(event_markers));
    }

    auto channel_count = file.channel_count();
    if (channel_count != 2) {
        throw std::runtime_error("expected 2 channels, got " + std::to_string(channel_count));
    }
    for (std::size_t i = 0; i < channel_count; i++) {
        auto channel = file.channel(i);
        if (channel.name == "accel") {
            check(channel.is_vector() &&
                      channel.components == std::vector<std::string>{"x", "y", "z"},
                  "accel components wrong");
        }
        if (channel.name == "heading_raw") {
            check(channel.period.has_value() && channel.period->as_degrees() == 360.0,
                  "heading_raw period wrong");
        }
    }
}

} // namespace

int main(int argc, char **argv) {
    try {
        const std::vector<std::string> args(argv, std::next(argv, argc));
        const fs::path base = args.size() >= 2 ? args.at(1) : "tests/fixtures/gold_dataset";
        const fs::path out = base / "gold_cpp.gtd";

        geotrace::FileBuilder builder{};
        builder.lenient();
        load_meta(builder, base);
        load_event_styles(builder, base);
        auto sats = load_satellites(base);
        load_fixes(builder, base, sats);
        load_markers(builder, base);
        load_events(builder, base);
        load_channels(builder, base);

        auto nav = builder.finish();
        nav.write_to_file(out);

        verify_counts(nav);

        std::cout << "Written: " << out << "\n";
        std::cout << "Gold dataset verified. Nav points: 200, Event markers: 7, Channels: 2\n";
    } catch (const std::exception &e) {
        std::cerr << "error: " << e.what() << "\n";
        return 1;
    }
}

File: sdk/python/geotrace-py/examples/gold_dataset.py
#!/usr/bin/env python3
"""Gold dataset reference example for the GeoTrace Python SDK.

Reads the CSV fixtures in ``tests/fixtures/gold_dataset/`` (the cross-SDK
reference data, mirrored by the Rust/C/C++ ``gold_dataset`` examples), builds a
``.gtd`` file, writes it as ``gold_py.gtd``, and verifies the round-trip.

Run it from anywhere - the repository root is located relative to this file.
"""

from __future__ import annotations

import csv
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path

from geotrace_sdk import (
    Annotation,
    Channel,
    Constellation,
    EventMarker,
    EventMarkerStyle,
    Marker,
    MarkerIcon,
    Meta,
    NavFileBuilder,
    NavFix,
    Satellite,
    SatelliteReport,
    TravelMode,
)

_ICONS = {
    "pin": MarkerIcon.PIN,
    "cross": MarkerIcon.CROSS,
    "circle": MarkerIcon.CIRCLE,
    "lightning": MarkerIcon.LIGHTNING,
    "warning": MarkerIcon.WARNING,
    "error": MarkerIcon.ERROR,
    "check": MarkerIcon.CHECK,
    "satellite": MarkerIcon.SATELLITE,
    "satellite_lost": MarkerIcon.SATELLITE_LOST,
    "gear": MarkerIcon.GEAR,
    "refresh": MarkerIcon.REFRESH,
    "download": MarkerIcon.DOWNLOAD,
    "upload": MarkerIcon.UPLOAD,
    "wrench": MarkerIcon.WRENCH,
}

_CONSTELLATIONS = {
    "gps": Constellation.GPS,
    "glonass": Constellation.GLONASS,
    "galileo": Constellation.GALILEO,
    "beidou": Constellation.BEIDOU,
}

# (`gps_time`, `sys_time`) raw CSV strings -> satellites captured at that instant.
SatKey = tuple[str, str]


def _find_repo_root() -> Path:
    """Walk up from this file until a directory holding the gold fixtures."""
    for parent in Path(__file__).resolve().parents:
        if (parent / "tests" / "fixtures" / "gold_dataset").is_dir():
            return parent
    msg = "could not locate tests/fixtures/gold_dataset above this file"
    raise FileNotFoundError(msg)


def _parse_ts(value: str) -> datetime | None:
    return datetime.fromisoformat(value) if value else None


def _opt_float(value: str) -> float | None:
    return float(value) if value else None


def _icon(name: str) -> MarkerIcon | None:
    return _ICONS.get(name)


def _rows(path: Path) -> list[list[str]]:
    """Return the data rows of a CSV file (header skipped), fields stripped."""
    # The fixtures contain UTF-8 (emoji in meta.csv), so be explicit to stop Windows
    # falling back to its cp1252 locale default and failing to decode them.
    with path.open(newline="", encoding="utf-8") as handle:
        rows = [[cell.strip() for cell in row] for row in csv.reader(handle)]
    return [row for row in rows[1:] if row and any(row)]


def _load_meta(base: Path) -> Meta:
    cols = _rows(base / "meta.csv")[0]
    return Meta(
        title=cols[0],
        device=cols[1],
        notes=cols[2],
        identity=cols[3],
        travel_mode=cols[4],
    )


def _load_event_styles(builder: NavFileBuilder, base: Path) -> None:
    for cols in _rows(base / "event_styles.csv"):
        builder.add_event_marker_style(
            EventMarkerStyle(
                cols[0],
                icon=_icon(cols[1]),
                color=cols[2] or None,
            )
        )


def _load_satellites(base: Path) -> dict[SatKey, list[Satellite]]:
    reports: dict[SatKey, list[Satellite]] = {}
    for cols in _rows(base / "satellites.csv"):
        sat = Satellite(
            _CONSTELLATIONS[cols[2]],
            int(cols[3]),
            in_fix=cols[4] == "true",
            elevation=_opt_float(cols[5]),
            azimuth=_opt_float(cols[6]),
            snr=_opt_float(cols[7]),
        )
        reports.setdefault((cols[0], cols[1]), []).append(sat)
    return reports


def _load_fixes(
    builder: NavFileBuilder,
    base: Path,
    satellites: dict[SatKey, list[Satellite]],
) -> None:
    for cols in _rows(base / "fixes.csv"):
        gps_time = _parse_ts(cols[1])
        sys_time = _parse_ts(cols[2])
        speed_kmh = _opt_float(cols[6])
        builder.add(
            NavFix(
                lat=float(cols[3]),
                lon=float(cols[4]),
                gps_time=gps_time,
                sys_time=sys_time,
                heading=_opt_float(cols[5]),
                # Match the SDK's MPS_PER_KMH = 1.0 / 3.6 constant-multiply so
                # the m/s value is bit-identical across SDKs (`kmh / 3.6` differs
                # by 1 ULP for some values).
                speed_mps=None if speed_kmh is None else speed_kmh * (1.0 / 3.6),
                eph_m=_opt_float(cols[7]),
            )
        )

        tracked = satellites.pop((cols[1], cols[2]), None)
        if tracked:
            builder.add(SatelliteReport(tracked, gps_time=gps_time, sys_time=sys_time))

    # Reports at a time no fix row holds. The builder gives each one a ghost fix.
    for (gps_col, sys_col), tracked in satellites.items():
        builder.add(
            SatelliteReport(
                tracked, gps_time=_parse_ts(gps_col), sys_time=_parse_ts(sys_col)
            )
        )


def _load_markers(builder: NavFileBuilder, base: Path) -> None:
    for cols in _rows(base / "markers.csv"):
        time = _parse_ts(cols[0])
        if time is None:
            continue
        icon = _icon(cols[2]) or MarkerIcon.PIN
        builder.add(Annotation(time=time, label=cols[1] or None, icon=icon))


def _load_events(builder: NavFileBuilder, base: Path) -> None:
    for cols in _rows(base / "events.csv"):
        time = _parse_ts(cols[0])
        if time is None:
            continue
        builder.add(EventMarker(cols[1], time, annotation=cols[2] or None))


@dataclass
class _ChannelAcc:
    """Accumulates one channel's metadata and samples across its CSV rows."""

    unit: str | None
    period_deg: float | None
    description: str | None
    components: list[str] | None
    times: list[datetime] = field(default_factory=list)
    values: list[float] = field(default_factory=list)


def _load_channels(builder: NavFileBuilder, base: Path) -> None:
    accumulators: dict[str, _ChannelAcc] = {}
    order: list[str] = []
    for cols in _rows(base / "channels.csv"):
        # cols: `name`, `unit`, `period_deg`, `description`, `components`,
        # `time`, `values`
        if len(cols) < 7:
            continue
        name = cols[0]
        acc = accumulators.get(name)
        if acc is None:
            acc = _ChannelAcc(
                unit=cols[1] or None,
                period_deg=float(cols[2]) if cols[2] else None,
                description=cols[3] or None,
                components=cols[4].split(";") if cols[4] else None,
            )
            accumulators[name] = acc
            order.append(name)
        time = _parse_ts(cols[5])
        if time is None:
            raise ValueError("channels.csv: invalid timestamp")
        acc.times.append(time)
        acc.values.extend(float(v) for v in cols[6].split(";"))

    for name in order:
        acc = accumulators[name]
        builder.add(
            Channel(
                name,
                acc.times,
                acc.values,
                unit=acc.unit,
                period_deg=acc.period_deg,
                description=acc.description,
                components=acc.components,
            )
        )


def _verify(path: Path) -> None:
    from geotrace_sdk import NavFile

    file = NavFile.open(path)

    meta = file.meta
    assert meta.title is not None and "Gold Dataset 🏆" in meta.title
    assert meta.device is not None and "Synthetic Generator 🧬" in meta.device
    assert meta.notes is not None and "🛰️" in meta.notes
    assert meta.travel_mode == TravelMode.BICYCLE

    points = file.points
    assert len(points) == 200, f"expected 200 nav points, got {len(points)}"

    antimeridian = [p for p in points if p.lon > 179.9 or p.lon < -179.9]
    assert len(antimeridian) == 11, f"antimeridian points: {len(antimeridian)}"

    stationary = [p for p in points if abs(p.lat - (-10.0)) < 1e-6]
    assert len(stationary) == 20, f"stationary points: {len(stationary)}"
    assert all(p.speed_mps == 0.0 for p in stationary)

    markers: list[Marker] = file.markers
    assert len(markers) == 16, f"expected 16 markers, got {len(markers)}"
    assert markers[0].label == "File Boundary Start"
    assert markers[0].icon == MarkerIcon.CHECK

    events = file.event_markers
    assert len(events) == 7, f"expected 7 event markers, got {len(events)}"

    styles = {s.variant_path: s for s in file.event_marker_styles}
    assert styles["style/custom-icon"].icon == MarkerIcon.LIGHTNING
    assert styles["style/custom-color"].color == "#FF00FF"

    channels = {c.name: c for c in file.channels}
    assert len(channels) == 2, f"expected 2 channels, got {len(channels)}"
    assert channels["accel"].is_vector
    assert channels["accel"].components == ["x", "y", "z"]
    assert channels["heading_raw"].period_deg == 360.0


def main() -> None:
    base = _find_repo_root() / "tests" / "fixtures" / "gold_dataset"

    builder = NavFileBuilder().with_meta(_load_meta(base))
    _load_event_styles(builder, base)
    _load_fixes(builder, base, _load_satellites(base))
    _load_markers(builder, base)
    _load_events(builder, base)
    _load_channels(builder, base)

    nav_file = builder.finish()

    out = base / "gold_py.gtd"
    nav_file.write_to_file(out)

    print(f"Gold dataset generated successfully: {out}")
    print(f"  Nav points:    {len(nav_file.points)}")
    print(f"  Markers:       {len(nav_file.markers)}")
    print(f"  Event markers: {len(nav_file.event_markers)}")
    print(f"  Channels:      {len(nav_file.channels)}")

    print("Verifying round-trip integrity...")
    _verify(out)
    print("Verified!")


if __name__ == "__main__":
    main()
